src/engine.rs

```diff
@@ -853,3 +858,194 @@ impl ResourcePool {
         }
     }
 }
+
+impl BindMapBuffer {
+    fn upload_if_needed(
```

Comment:
Am I understanding this correctly, that once this gets called a materialized CPU buffer gets converted to a GPU buffer and can no longer be used in a CPU pipeline (since `cpu_buf` gets overwritten in the assignment below to `self.buffer = MaterializedBuffer::Gpu(buf);`?

I think the function name should reflect this conversion (or "re-materialization") as it's more than just uploading data from the CPU to the GPU.

Reply:
I added a comment but was not inspired with a better name. It's not part of the public API.

Reply:
I think the comment looks good and explains the side-effect that I wanted to captured.